x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/CategorizeInternal.java

```diff
@@ -0,0 +1,149 @@
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.ml.aggs.categorization.CategorizationBytesRefHash;
+import org.elasticsearch.xpack.ml.aggs.categorization.CategorizationPartOfSpeechDictionary;
+import org.elasticsearch.xpack.ml.aggs.categorization.TokenListCategorizer.CloseableTokenListCategorizer;
+import org.elasticsearch.xpack.ml.job.categorization.CategorizationAnalyzer;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+
+/**
+ * Categorizes text messages; is used internally in the Categorize grouping function.
+ */
+public class CategorizeInternal extends EsqlScalarFunction {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "CategorizeInternal",
+        CategorizeInternal::new
+    );
+
+    private final Expression str;
+
+    @FunctionInfo(returnType = "integer", description = "Categorizes text messages")
+    public CategorizeInternal(
+        Source source,
+        @Param(name = "string", type = { "keyword", "text" }, description = "String expression.") Expression str
+    ) {
+        super(source, Collections.singletonList(str));
+        this.str = str;
+    }
+
+    private CategorizeInternal(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(str);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Evaluator
+    static int process(
+        BytesRef v,
+        @Fixed(includeInToString = false, build = true) CategorizationAnalyzer analyzer,
+        @Fixed(includeInToString = false, build = true) CloseableTokenListCategorizer categorizer
+    ) {
+        String s = v.utf8ToString();
+        try (TokenStream ts = analyzer.tokenStream("text", s)) {
+            return categorizer.computeCategory(ts, s.length(), 1).getId();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        return new CategorizeInternalEvaluator.Factory(
+            source(),
+            toEvaluator.apply(str),
+            context -> new CategorizationAnalyzer(
+                // TODO(jan): get the correct analyzer in here, see CategorizationAnalyzerConfig::buildStandardCategorizationAnalyzer
```

Comment:
So I understand, is this a "must" before this function works, or is it a good to have? Will be done here?

Reply:
This should be resolved before going live with this. I might fix it here. Right now, this implementation is an incomplete snapshot function anyway, so it's not really important.

Reply:
Do you have any advise on how to best get analyzers, char filters and token filters in here?

Reply:
Oh boy! I don't know to be honest. Analyzers are usually a concept bound to an index, but ESQL queries are cross-index. But this analyzer is generic, right?

Reply:
Yes, it's a generic one.

Something like:
- tokenizer: ml_standard
- char filter: first_line_with_letters
- token filter: stopword (with custom stopword list)
- token filter: max token count

Reply:
Note for the future: the [Categorize text aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-categorize-text-aggregation.html) allows you to specify a `categorization_analyzer` in the request body. I think we want that too in ES|QL eventually, but not needed for the first version.